### PR TITLE
feat: Internal cleanup for buffered/queued messages on failure/cancel

### PR DESCRIPTION
**Pull request by johanandren**
Alternative or complementary to https://github.com/akka/akka-http/pull/4299 for https://github.com/akka/akka-grpc/issues/1423

Allows for failing buffered requests if the persistent http client fails. Still not water tight for the general case but maybe we can make water tight for Akka HTTP/2 / gRPC, might not want to make it public API for that reason.

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala [hunk 1]
```diff
@@ -223,6 +229,16 @@ import akka.stream.stage._
 
       override def fail(ex: Throwable): Unit = callback.invoke(Failure(ex))
 
+      private def handleBufferedOnFailure(ex: Throwable): Unit = {
+        if (buffer != null && buffer.nonEmpty) {
+          onBufferedFailureOrCancel match {
+            case Some(f) =>
+              while (buffer.nonEmpty) f(ex, buffer.dequeue())
+            case None =>
+              log.debug("Queue source failing, [{}] elements in buffer dropped", buffer.used)
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala:238]: Maybe such events should be fed to dead letters instead to align with other places messages are dropped?

Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala:238]: do we have that in any other Akka Streams buffers?

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala:238]: Nope, can't see we do that anywhere in streams, we probably started doing that with typed. Let's not then.

**Comment by johanandren** [comment]
I have a new idea about solution for the gRPC problem, so let's close this for now

**Pull request closed without merging**